File: src/solvers/robust_pomcp.jl
"""
Solver A: Robust-Value POMCP

Builds a search tree via Monte Carlo simulations (nominal model),
then computes robust values via worst-case Bellman backups using
projected uncertainty sets.

Tunable parameters:
  sims_per_backup: number of simulations between successive robust backups.
    sims_per_backup == budget is two-phase; sims_per_backup == 1 is fully interleaved.
  ucb_mode: :nominal (optimistic) or :robust (pessimistic) exploration

Observability:
  Pass a SolverLog to `logger` and the solver emits fine-grained events
  (sim_start, sim_step, expand, rollout, backup_nominal_step, sim_end,
  robust_backup) in chronological order for later replay.
"""

# --- Top-level planner ---

"""
    robust_pomcp_plan(belief, model, uncertainty, horizon, budget, sims_per_backup;
                      ucb_mode, exploration_constant, logger)

Plan a single action using Robust-Value POMCP.

`sims_per_backup` controls how many simulations run between successive robust
backups. `sims_per_backup == budget` is two-phase; `sims_per_backup == 1` is
fully interleaved.

Returns: (best_action, root)
"""
function robust_pomcp_plan(belief::Vector{Float64}, model::TabularPOMDP,
                           uncertainty::UncertaintySets, horizon::Int,
                           budget::Int, sims_per_backup::Int;
                           ucb_mode::Symbol=:nominal,
                           exploration_constant::Float64=1.0,
                           logger::Union{SolverLog, Nothing}=nothing)

    id_counter = NodeIdCounter()
    root_id = next_node_id!(id_counter)
    root = HistoryNode(root_id, 0)

    # State machinery for robust_backup indexing across the run
    backup_counter = Ref(0)

    for batch in 1:ceil(Int, budget / sims_per_backup)
        n_sims_this_batch = min(sims_per_backup, budget - (batch - 1) * sims_per_backup)

        for _ in 1:n_sims_this_batch
            sim_index = _next_sim_index(logger)
            s = sample_from_belief(belief)

            if logger !== nothing
                emit_sim_start!(logger, sim_index, s, root.id)
            end

            step_counter = Ref(0)
            total_ret = simulate!(s, root, horizon, model, ucb_mode,
                                  exploration_constant, id_counter,
                                  logger, sim_index, step_counter)

            if logger !== nothing
                emit_sim_end!(logger; sim_index=sim_index, total_return=total_ret)
            end
        end

        robust_backup!(root, model, uncertainty, logger, backup_counter)
    end

    # Pick action with highest Q_robust
    best_action = 0
    best_value = -Inf
    for (a, action_node) in root.children
        if action_node.Q_robust > best_value
            best_value = action_node.Q_robust
            best_action = a
        end
    end

    return (best_action, root)
end

"""Return the next sim_index by counting sim_start events already emitted."""
function _next_sim_index(logger::Union{SolverLog, Nothing})
    logger === nothing && return 0
    count = 0
    for ev in logger.events
        ev.type === :sim_start && (count += 1)
    end
    return count + 1
end

# --- Simulation ---

"""
Run one POMCP simulation from state s at the given history node.

If a `logger` is present, emits sim_step, expand, rollout, and backup_nominal_step
events in execution order.

Returns: total return from this point downward.
"""
function simulate!(s::Int, 
                   node::HistoryNode, 
                   depth::Int, 
                   model::TabularPOMDP,
                   ucb_mode::Symbol, 
                   c::Float64,
                   id_counter::NodeIdCounter,
                   logger::Union{SolverLog, Nothing},
                   sim_index::Int,
                   step_counter::Ref{Int})
    if depth == 0
        return 0.0
    end

    add_particle!(node, s)

    # Leaf node: expand and rollout
    if is_leaf(node)
        created_ids, created_actions = expand!(node, model.n_actions, id_counter)
        if logger !== nothing
            emit_expand!(logger;
                sim_index=sim_index, node_id=node.id, depth=node.depth,
                created_action_nodes=created_ids,
                created_action_indices=created_actions)
        end

        # Rollout
        rollout_return, rollout_steps = rollout_with_log(s, depth, model)
        node.V_rollout = rollout_return
        node.has_rollout = true

        if logger !== nothing
            emit_rollout!(logger;
                sim_index=sim_index,
                leaf_node_id=node.id,
                start_state=s,
                start_depth=node.depth,
                rollout_steps=rollout_steps,
                total_return=rollout_return)
        end

        return rollout_return
    end

    # Select action via UCB — capture the UCB values we actually saw
    a, ucb_qs, ucb_ns = ucb_select_with_info(node, ucb_mode, c, model.n_actions)

    # Sample next state, observation, reward (under nominal model)
    s_next = sample_transition(model, s, a)
    z = sample_observation(model, s_next)
    r = reward(model, s, a)

    # Get or create action/obs children; track newly-created ids
    action_node = node.children[a]
    child, child_was_created = get_or_create_child!(action_node, z,
                                                    node.depth + 1, id_counter)

    if logger !== nothing
        step_counter[] += 1
        emit_sim_step!(logger;
            sim_index=sim_index,
            step_in_sim=step_counter[],
            active_history_id=node.id,
            depth=node.depth,
            state=s,
            ucb_qs=ucb_qs,
            ucb_ns=ucb_ns,
            action_chosen=a,
            action_node_id=action_node.id,
            s_next=s_next,
            obs=z,
            reward=r,
            next_history_id=child.id,
            created_action_nodes=Int[],
            created_history_nodes=child_was_created ? [child.id] : Int[])
    end

    # Recurse
    future_return = simulate!(s_next, child, depth - 1, model, ucb_mode, c,
                              id_counter, logger, sim_index, step_counter)
    total_return = r + future_return

    # Update Q_nominal (incremental mean) — emit a backup_nominal_step event
    prev_Q = action_node.Q_nominal
    prev_N = action_node.N
    action_node.N += 1
    action_node.Q_nominal += (total_return - action_node.Q_nominal) / action_node.N
    action_node.dirty = true

    if logger !== nothing
        emit_backup_nominal_step!(logger;
            sim_index=sim_index,
            action_node_id=action_node.id,
            previous_Q_nominal=prev_Q,
            new_Q_nominal=action_node.Q_nominal,
            previous_N=prev_N,
            new_N=action_node.N,
            total_return_from_here=total_return)
    end

    update_nominal_value!(node)
    return total_return
end

# --- Rollout ---

"""Random rollout. Returns (total_return, steps) where steps is a list of dicts."""
function rollout_with_log(s::Int, depth::Int, model::TabularPOMDP)
    total_return = 0.0
    current_s = s
    steps = Dict{String, Any}[]

    for _ in 1:depth
        a = rand(1:model.n_actions)
        r = reward(model, current_s, a)
        s_next = sample_transition(model, current_s, a)
        push!(steps, Dict{String, Any}(
            "state"  => current_s,
            "action" => a,
            "reward" => r,
            "s_next" => s_next,
        ))
        total_return += r
        current_s = s_next
    end

    return (total_return, steps)
end

# --- UCB Action Selection ---

"""
UCB selection that also returns the Q values and visit counts used for the
decision (for logging). Action order is 1..n_actions.
"""
function ucb_select_with_info(node::HistoryNode, ucb_mode::Symbol, c::Float64,
                              n_actions::Int)
    ucb_qs = Float64[]
    ucb_ns = Int[]

    for a in 1:n_actions
        an = node.children[a]
        q = (ucb_mode == :robust) ? an.Q_robust : an.Q_nominal
        push!(ucb_qs, q)
        push!(ucb_ns, an.N)
    end

    # Visit unvisited actions first
    for a in 1:n_actions
        if node.children[a].N == 0
            return (a, ucb_qs, ucb_ns)
        end
    end

    # Standard UCB
    log_N = log(node.N)
    best_action = 0
    best_ucb = -Inf
    for a in 1:n_actions
        an = node.children[a]
        q = (ucb_mode == :robust) ? an.Q_robust : an.Q_nominal
        ucb_value = q + c * sqrt(log_N / an.N)
        if ucb_value > best_ucb
            best_ucb = ucb_value
            best_action = a
        end
    end

    return (best_action, ucb_qs, ucb_ns)
end

# --- Robust Backup ---

"""
Bottom-up robust value computation. Only recomputes dirty nodes.

Invariant exploited here: if a node is not dirty, NONE of its descendants are
either. So we short-circuit the entire subtree at the first clean node hit.
This saves significant work when K < N (interleaved backups on large trees).
"""
function robust_backup!(node::HistoryNode, model::TabularPOMDP,
                        uncertainty::UncertaintySets,
                        logger::Union{SolverLog, Nothing},
                        backup_counter::Ref{Int})
    # Short-circuit: clean subtree means no work below this node.
    if !node.dirty
        return
    end

    for (a, action_node) in node.children
        # Skip entire action subtree if the action node itself is clean —
        # its Q_robust is still correct from a previous backup, and its
        # descendants can't be dirty (see invariant above).
        if !action_node.dirty
            continue
        end

        # Recurse into observation children (bottom-up). Individual clean
        # subtrees bail out at their entry via the node.dirty check above.
        for (_, child) in action_node.children
            robust_backup!(child::HistoryNode, model, uncertainty, logger, backup_counter)
        end

        prev_Qr = action_node.Q_robust
        if isempty(action_node.children)
            action_node.Q_robust = action_node.Q_nominal
        else
            action_node.Q_robust = compute_robust_q(node, a, action_node, model,
                                                   uncertainty, logger,
                                                   backup_counter, prev_Qr)
        end
        action_node.dirty = false
    end

    update_robust_value!(node)
    node.dirty = false
end

# --- Core: Robust Q Computation (two-step LP decomposition) ---

function compute_robust_q(h::HistoryNode, a::Int, action_node::ActionNode,
                          model::TabularPOMDP, uncertainty::UncertaintySets,
                          logger::Union{SolverLog, Nothing},
                          backup_counter::Ref{Int},
                          previous_Q_robust::Float64)
    S_in = collect(h.S_in)
    Z_in = collect(action_node.Z_in)

    belief = compute_belief(h.particles, S_in)

    V_children = Float64[
        haskey(action_node.children, z) ?
            (action_node.children[z]::HistoryNode).V_robust : 0.0
        for z in Z_in
    ]

    obs_lp_results = ObsLPResult[]
    trans_lp_results = TransLPResult[]

    # Step 1: observation LP per s'
    w = Vector{Float64}(undef, length(S_in))
    for (i, s_next) in enumerate(S_in)
        p_nominal_obs = Float64[observation_prob(model, s_next, z) for z in Z_in]
        ρ_z = observation_radius(uncertainty, s_next)
        w_val, optimal_p = minimize_over_projected_tv_ball(p_nominal_obs, ρ_z, V_children)
        w[i] = w_val
        if logger !== nothing
            push!(obs_lp_results, ObsLPResult(s_next, p_nominal_obs, ρ_z, optimal_p, w_val))
        end
    end

    # Step 2: transition LP per s
    Q_val = 0.0
    for (i, s) in enumerate(S_in)
        p_nominal_trans = Float64[transition_prob(model, s, a, s_next) for s_next in S_in]
        ρ_t = transition_radius(uncertainty, s, a)
        σ, optimal_p = minimize_over_projected_tv_ball(p_nominal_trans, ρ_t, w)
        Q_val += belief[i] * (reward(model, s, a) + σ)
        if logger !== nothing
            push!(trans_lp_results, TransLPResult(s, p_nominal_trans, ρ_t, optimal_p, σ))
        end
    end

    if logger !== nothing
        backup_counter[] += 1
        emit_robust_backup!(logger;
            backup_index=backup_counter[],
            history_node_id=h.id,
            action_node_id=action_node.id,
            node_depth=h.depth,
            action=a,
            S_in=S_in,
            Z_in=Z_in,
            belief=belief,
            V_children=V_children,
            obs_lp_results=obs_lp_results,
            trans_lp_results=trans_lp_results,
            previous_Q_robust=previous_Q_robust,
            new_Q_robust=Q_val)
    end

    return Q_val
end

# --- Helpers ---

"""Compute empirical belief from particles over S_in. Returns vector aligned with S_in."""
function compute_belief(particles::Vector{Int}, S_in::Vector{Int})
    counts = zeros(Int, length(S_in))
    state_to_idx = Dict(s => i for (i, s) in enumerate(S_in))
    for s in particles
        if haskey(state_to_idx, s)
            counts[state_to_idx[s]] += 1
        end
    end
    total = sum(counts)
    return counts ./ total
end

"""Sample a state from a belief vector."""
function sample_from_belief(belief::Vector{Float64})
    return rand(Categorical(belief))
end

# --- Diagnostics ---

"""Count HistoryNodes and ActionNodes in a tree."""
function tree_size(root::HistoryNode)
    n_history = 1
    n_action = 0
    for (_, action_node) in root.children
        n_action += 1
        for (_, child) in action_node.children
            nh, na = tree_size(child::HistoryNode)
            n_history += nh
            n_action += na
        end
    end
    return (n_history, n_action)
end

File: src/robust_pomdp/solvers/robust_pomcp.py
"""
Solver A: Robust-Value POMCP.

Builds a search tree via Monte Carlo simulations (under the nominal model),
then computes robust values via worst-case Bellman backups using projected
uncertainty sets.

Tunable parameters:
    sims_per_backup: simulations between successive robust backups.
        sims_per_backup == budget is two-phase; sims_per_backup == 1 is fully
        interleaved.
    ucb_mode: "nominal" (optimistic) or "robust" (pessimistic) exploration.

Observability:
    Pass a SolverLog as `logger` and the solver emits fine-grained events
    (sim_start, sim_step, expand, rollout, backup_nominal_step, sim_end,
    robust_backup) in chronological order for later replay.

Preserved Julia invariants (Sessions 5–7):
    - V_rollout fallback for unvisited branches (HistoryNode methods).
    - V_nominal repropagation after every Q_nominal update in simulate.
    - Dirty-flag invariant: clean node => no descendant is dirty. Backup
      short-circuits at the first clean node hit.
    - HistoryNode.N bumped on every visit (in add_particle).
"""

from __future__ import annotations

from typing import Any, Literal

import numpy as np

from robust_pomdp.core.pomdp_types import TabularPOMDP
from robust_pomdp.core.tree import ActionNode, HistoryNode, NodeIdCounter
from robust_pomdp.debug.logger import ObsLPResult, SolverLog, TransLPResult
from robust_pomdp.uncertainty.projected_sets import ProjectedTVBall
from robust_pomdp.uncertainty.uncertainty_sets import UncertaintySets


UCBMode = Literal["nominal", "robust"]


# ---------------------------------------------------------------------------
# Top-level planner
# ---------------------------------------------------------------------------

def robust_pomcp_plan(belief: np.ndarray,
                      model: TabularPOMDP,
                      uncertainty: UncertaintySets,
                      horizon: int,
                      budget: int,
                      sims_per_backup: int,
                      *,
                      ucb_mode: UCBMode = "nominal",
                      exploration_constant: float = 1.0,
                      logger: SolverLog | None = None,
                      rng: np.random.Generator | None = None
                      ) -> tuple[int, HistoryNode]:
    """Plan a single action with Robust-Value POMCP.

    Returns:
        (best_action, root) — best_action is 0-based.
    """
    if rng is None:
        rng = np.random.default_rng()
    belief = np.asarray(belief, dtype=np.float64)

    id_counter = NodeIdCounter()
    root = HistoryNode(id_counter.next_id(), depth=0)

    # LP reuse: cache ProjectedTVBall instances by problem size n. The LP
    # structure depends only on n; (p_nominal, rho, c) are passed per solve.
    lp_cache: dict[int, ProjectedTVBall] = {}

    backup_counter = 0
    n_batches = -(-budget // sims_per_backup)  # ceil(budget / sims_per_backup)

    for batch in range(n_batches):
        n_sims_this_batch = min(sims_per_backup, budget - batch * sims_per_backup)

        for _ in range(n_sims_this_batch):
            sim_index = _next_sim_index(logger)
            s = sample_from_belief(belief, rng)

            if logger is not None:
                logger.emit_sim_start(sim_index=sim_index, initial_state=s, root_node_id=root.id,)

            total_ret, _ = _simulate(s,
                                     root,
                                     horizon,
                                     model,
                                     ucb_mode,
                                     exploration_constant,
                                     id_counter,
                                     logger,
                                     sim_index,
                                     step_counter=0,
                                     rng=rng
                                     )

            if logger is not None:
                logger.emit_sim_end(sim_index=sim_index, total_return=total_ret)

        backup_counter = _robust_backup(root, model, uncertainty, logger, backup_counter, lp_cache,)

    # Pick action with highest Q_robust at the root.
    best_action = -1
    best_value = -np.inf
    for a, action_node in root.children.items():
        if action_node.Q_robust > best_value:
            best_value = action_node.Q_robust
            best_action = a

    return best_action, root


def _next_sim_index(logger: SolverLog | None) -> int:
    """Return 1 + (number of sim_start events emitted so far). 0 if no logger."""
    if logger is None:
        return 0
    count = sum(1 for ev in logger.events if ev.type == "sim_start")
    return count + 1


# ---------------------------------------------------------------------------
# Simulation
# ---------------------------------------------------------------------------

def _simulate(s: int,
              node: HistoryNode,
              depth: int,
              model: TabularPOMDP,
              ucb_mode: UCBMode,
              c: float,
              id_counter: NodeIdCounter,
              logger: SolverLog | None,
              sim_index: int,
              step_counter: int,
              rng: np.random.Generator
              ) -> tuple[float, int]:
    """One POMCP simulation from state s at the given history node.

    Returns (total_return_from_here, updated_step_counter).
    """
    if depth == 0:
        return 0.0, step_counter

    node.add_particle(s)

    # Leaf: expand and rollout.
    if node.is_leaf():
        created_ids, created_actions = node.expand(model.n_actions, id_counter)
        if logger is not None:
            logger.emit_expand(sim_index=sim_index,
                               node_id=node.id,
                               depth=node.depth,
                               created_action_nodes=created_ids,
                               created_action_indices=created_actions
                               )

        rollout_return, rollout_steps = _rollout_with_log(s, depth, model, rng)
        node.V_rollout = rollout_return
        node.has_rollout = True

        if logger is not None:
            logger.emit_rollout(sim_index=sim_index,
                                leaf_node_id=node.id,
                                start_state=s,
                                start_depth=node.depth,
                                rollout_steps=rollout_steps,
                                total_return=rollout_return
                                )

        return rollout_return, step_counter

    # Select action via UCB.
    a, ucb_qs, ucb_ns = _ucb_select_with_info(node, ucb_mode, c, model.n_actions)

    # Sample next state, observation, reward (under nominal model).
    s_next = model.sample_transition(s, a, rng)
    z = model.sample_observation(s_next, rng)
    r = model.reward(s, a)

    action_node = node.children[a]
    child, child_was_created = action_node.get_or_create_child(z, node.depth + 1, id_counter)

    if logger is not None:
        step_counter += 1
        logger.emit_sim_step(sim_index=sim_index,
                             step_in_sim=step_counter,
                             active_history_id=node.id,
                             depth=node.depth,
                             state=s,
                             ucb_qs=ucb_qs,
                             ucb_ns=ucb_ns,
                             action_chosen=a,
                             action_node_id=action_node.id,
                             s_next=s_next,
                             obs=z,
                             reward=r,
                             next_history_id=child.id,
                             created_action_nodes=[],
                             created_history_nodes=[child.id] if child_was_created else []
                             )

    # Recurse.
    future_return, step_counter = _simulate(s_next,
                                            child,
                                            depth - 1,
                                            model,
                                            ucb_mode,
                                            c,
                                            id_counter,
                                            logger,
                                            sim_index,
                                            step_counter,
                                            rng
                                            )
    total_return = r + future_return

    # Update Q_nominal (incremental mean) and dirty the action node.
    prev_Q = action_node.Q_nominal
    prev_N = action_node.N
    action_node.N += 1
    action_node.Q_nominal += (
        (total_return - action_node.Q_nominal) / action_node.N
    )
    action_node.dirty = True

    if logger is not None:
        logger.emit_backup_nominal_step(sim_index=sim_index,
                                        action_node_id=action_node.id,
                                        previous_Q_nominal=prev_Q,
                                        new_Q_nominal=action_node.Q_nominal,
                                        previous_N=prev_N,
                                        new_N=action_node.N,
                                        total_return_from_here=total_return
                                        )

    # V_nominal of `node` must be recomputed from its children because the
    # max-Q child may not be the one we just updated (Session 6 fix).
    node.update_nominal_value()

    return total_return, step_counter


# ---------------------------------------------------------------------------
# Rollout
# ---------------------------------------------------------------------------

def _rollout_with_log(s: int,
                      depth: int,
                      model: TabularPOMDP,
                      rng: np.random.Generator
                      ) -> tuple[float, list[dict[str, Any]]]:
    """Random rollout. Returns (total_return, steps_list)."""
    total_return = 0.0
    current_s = s
    steps: list[dict[str, Any]] = []

    for _ in range(depth):
        a = int(rng.integers(model.n_actions))
        r = model.reward(current_s, a)
        s_next = model.sample_transition(current_s, a, rng)
        steps.append({
            "state":  current_s,
            "action": a,
            "reward": r,
            "s_next": s_next,
        })
        total_return += r
        current_s = s_next

    return total_return, steps


# ---------------------------------------------------------------------------
# UCB action selection
# ---------------------------------------------------------------------------

def _ucb_select_with_info(node: HistoryNode,
                          ucb_mode: UCBMode,
                          c: float,
                          n_actions: int
                          ) -> tuple[int, list[float], list[int]]:
    """UCB selection that also returns the (Q values, visit counts) seen.

    Returns (best_action, ucb_qs, ucb_ns) where ucb_qs[a] / ucb_ns[a] are
    aligned to action index a in 0..n_actions-1.
    """
    ucb_qs: list[float] = []
    ucb_ns: list[int] = []

    for a in range(n_actions):
        an = node.children[a]
        q = an.Q_robust if ucb_mode == "robust" else an.Q_nominal
        ucb_qs.append(q)
        ucb_ns.append(an.N)

    # Visit unvisited actions first.
    for a in range(n_actions):
        if node.children[a].N == 0:
            return a, ucb_qs, ucb_ns

    # Standard UCB.
    log_N = np.log(node.N)
    best_action = -1
    best_ucb = -np.inf
    for a in range(n_actions):
        an = node.children[a]
        q = an.Q_robust if ucb_mode == "robust" else an.Q_nominal
        ucb_value = q + c * np.sqrt(log_N / an.N)
        if ucb_value > best_ucb:
            best_ucb = ucb_value
            best_action = a

    return best_action, ucb_qs, ucb_ns


# ---------------------------------------------------------------------------
# Robust backup (bottom-up, dirty-aware)
# ---------------------------------------------------------------------------

def _robust_backup(node: HistoryNode,
                   model: TabularPOMDP,
                   uncertainty: UncertaintySets,
                   logger: SolverLog | None,
                   backup_counter: int,
                   lp_cache: dict[int, ProjectedTVBall]
                   ) -> int:
    """Bottom-up robust value computation. Recomputes only dirty nodes.

    Invariant: clean node => no descendant is dirty. We short-circuit the
    entire subtree at the first clean node.

    Returns updated backup_counter.
    """
    if not node.dirty:
        return backup_counter

    for a, action_node in node.children.items():
        if not action_node.dirty:
            continue

        # Recurse into observation children first (bottom-up).
        for child in action_node.children.values():
            backup_counter = _robust_backup(child,
                                            model,
                                            uncertainty,
                                            logger,
                                            backup_counter,
                                            lp_cache
                                            )

        prev_Qr = action_node.Q_robust
        if not action_node.children:
            # No observation children yet — the action was visited at a leaf
            # but never followed up. Q_robust falls back to Q_nominal.
            action_node.Q_robust = action_node.Q_nominal
        else:
            new_Qr, backup_counter = compute_robust_q(node,
                                                       a,
                                                       action_node,
                                                       model,
                                                       uncertainty,
                                                       logger,
                                                       backup_counter,
                                                       prev_Qr,
                                                       lp_cache
                                                       )
            action_node.Q_robust = new_Qr

        action_node.dirty = False

    node.update_robust_value()
    node.dirty = False

    return backup_counter


# ---------------------------------------------------------------------------
# Two-step LP decomposition for robust Q
# ---------------------------------------------------------------------------

def compute_robust_q(h: HistoryNode,
                     a: int,
                     action_node: ActionNode,
                     model: TabularPOMDP,
                     uncertainty: UncertaintySets,
                     logger: SolverLog | None,
                     backup_counter: int,
                     previous_Q_robust: float,
                     lp_cache: dict[int, ProjectedTVBall] | None = None,
                     belief_override: np.ndarray | None = None
                     ) -> tuple[float, int]:
    """Compute Q_robust(h, a) via the two-step LP decomposition.

    Step 1 (per s'): worst-case observation distribution -> w(s').
    Step 2 (per s):  worst-case transition distribution -> σ(s).
    Q_val = Σ_s belief[s] · (R(s, a) + σ(s)).

    Returns (Q_val, updated_backup_counter).

    `lp_cache` may be None (e.g. when called from a verification script that
    only does one or two backups). A fresh dict is allocated in that case.

    `belief_override`, if given, replaces the particle-empirical belief with an
    explicit vector aligned with `sorted(h.S_in)`. Used by A1's exact tree
    evaluator (no Monte-Carlo). POMCP-style callers leave it as `None`.
    """
    if lp_cache is None:
        lp_cache = {}

    S_in = sorted(h.S_in)
    Z_in = sorted(action_node.Z_in)

    if belief_override is not None:
        assert belief_override.shape == (len(S_in),), \
            f"belief_override shape {belief_override.shape} != ({len(S_in)},)"
        belief = belief_override
    else:
        belief = compute_belief(h.particles, S_in)

    V_children = [action_node.children[z].V_robust if z in action_node.children else 0.0 for z in Z_in]

    obs_lp_results: list[ObsLPResult] = []
    trans_lp_results: list[TransLPResult] = []

    n_z = len(Z_in)
    n_s = len(S_in)
    full_support_obs = (n_z == model.n_obs)
    full_support_trans = (n_s == model.n_states)
    if n_z not in lp_cache:
        lp_cache[n_z] = ProjectedTVBall(n=n_z)
    if n_s not in lp_cache:
        lp_cache[n_s] = ProjectedTVBall(n=n_s)
    lp_obs = lp_cache[n_z]
    lp_trans = lp_cache[n_s]

    # Step 1: observation LP per s'.
    w = np.empty(n_s, dtype=np.float64)
    for i, s_next in enumerate(S_in):
        p_nominal_obs = [model.observation_prob(s_next, z) for z in Z_in]
        rho_z = uncertainty.observation_radius(s_next)
        w_val, optimal_p = lp_obs.solve(p_nominal_obs, rho_z, V_children,
                                        full_support=full_support_obs)
        w[i] = w_val
        if logger is not None:
            obs_lp_results.append(ObsLPResult(s_next=s_next,
                                              nominal_p=list(p_nominal_obs),
                                              radius=float(rho_z),
                                              optimal_p=optimal_p.tolist(),
                                              objective_value=float(w_val)
                                              ))

    # Step 2: transition LP per s.
    Q_val = 0.0
    for i, s in enumerate(S_in):
        p_nominal_trans = [model.transition_prob(s, a, s_next) for s_next in S_in]
        rho_t = uncertainty.transition_radius(s, a)
        sigma, optimal_p = lp_trans.solve(p_nominal_trans, rho_t, w.tolist(),
                                          full_support=full_support_trans)
        Q_val += belief[i] * (model.reward(s, a) + sigma)
        if logger is not None:
            trans_lp_results.append(TransLPResult(state=s,
                                                  nominal_p=list(p_nominal_trans),
                                                  radius=float(rho_t),
                                                  optimal_p=optimal_p.tolist(),
                                                  objective_value=float(sigma)
                                                  ))

    if logger is not None:
        backup_counter += 1
        logger.emit_robust_backup(backup_index=backup_counter,
                                  history_node_id=h.id,
                                  action_node_id=action_node.id,
                                  node_depth=h.depth,
                                  action=a,
                                  S_in=list(S_in),
                                  Z_in=list(Z_in),
                                  belief=belief.tolist(),
                                  V_children=list(V_children),
                                  obs_lp_results=obs_lp_results,
                                  trans_lp_results=trans_lp_results,
                                  previous_Q_robust=float(previous_Q_robust),
                                  new_Q_robust=float(Q_val)
                                  )

    return float(Q_val), backup_counter


# ---------------------------------------------------------------------------
# Helpers (public)
# ---------------------------------------------------------------------------

def compute_belief(particles: list[int], S_in: list[int]) -> np.ndarray:
    """Empirical belief over S_in from particles. Returns vector aligned with S_in."""
    counts = np.zeros(len(S_in), dtype=np.float64)
    state_to_idx = {s: i for i, s in enumerate(S_in)}
    for s in particles:
        if s in state_to_idx:
            counts[state_to_idx[s]] += 1.0
    total = counts.sum()
    if total == 0.0:
        return counts
    return counts / total


def sample_from_belief(belief: np.ndarray, rng: np.random.Generator) -> int:
    """Sample a state from a belief vector."""
    return int(rng.choice(len(belief), p=belief))


def tree_size(root: HistoryNode) -> tuple[int, int]:
    """Count (HistoryNodes, ActionNodes) in the tree rooted at `root`."""
    n_history = 1
    n_action = 0
    for action_node in root.children.values():
        n_action += 1
        for child in action_node.children.values():
            nh, na = tree_size(child)
            n_history += nh
            n_action += na
    return n_history, n_action
